### PR TITLE
[158] Add role-based permission check to routes

### LINKED ISSUE #13: Challenge manager role access
### User Story
As a user with a challenge manager role, in order to setup evaluations for the submissions of my challenges and efficiently manage the entire evaluation process, I would like to be able to access appropriate sections of the site upon login.

**Acceptance criteria:**

- [x] When a user with a challenge manager role logs in, they are sent to the Challenge Manager's dashboard on the Rails app, #53 
- [x] An appropriate utility menu is displayed for the user, https://github.com/GSA/Challenge_platform/issues/27
- [x] If any other user who does not have a challenge manager's role assigned to them tries to access challenge manager role specific features, they are redirected to their /dashboard (based on their assigned role) and an error message is displayed: "You do not have access to this section of Challenge.gov."

**Development tasks:**

- [x] #158

### Definition of Done

**Doing (dev team)**
- [x] Code complete 
- [x] Code is organized appropriately
- [x] Any known trade offs are documented in the associated GH issue
- [x] Code is documented, modules, shared functions, etc.
- [x] Automated testing has been added or updated in response to changes in this PR
- [x] The feature is smoke tested to confirm it meets requirements
- [x] Database changes have been peer reviewed for index changes and performance bottlenecks
- [ ] ~PR that changes or adds UI~
  - [ ] ~include a screenshot of the WAVE report for the altered pages~
  - [ ] ~Confirm changes were validated for mobile responsiveness~
- [x] PR approved / Peer reviewed 
- [x] Security scans passed
- [x] Automate accessibility tests passed
- [x] Build process and deployment is automated and repeatable
- [x] Feature toggles if appropriate
- [x] Deploy to staging
- [x] Move card to testing column in the board
---
**Staging**
- [x] Accessibility tested (Marni)
  - [x]  Keyboard navigation
  - [x]  Focus confirmed
  - [x] Color contrast compliance
  - [x] Screen reader testing 
- [x] Usability testing: mobile and desktop (Tracy or Marni)
- [x] Cross browser testing - UI rendering is performant on below listed devices/browsers (Tracy or Marni)
  - [x] Windows/Chrome
  - [x] Windows/Edge
  - [x] Mac/Chrome
  - [x] Mac/Safari
  - [x] iOS/Safari
- [x] AC review (Renata)
- [ ] ~Deploy to production (production-like environment for eval capability) (dev team)~
- [x] Move to production column in the board
---
**Production**
- [ ] User and security documentation has been reviewed for necessary updates (Renata/Tracy/Dev team) 
- [x] PO / PM approved (Jarah or Renata)
- [x] AC is met and it works as expected (Jarah or Renata)
- [ ] Move to done column in the board (Jarah or Renata)

### LINKED ISSUE #16: Evaluator role access
### User Story
As a user with an evaluator role, in order to evaluate submissions assigned to me, I would like to be able to access evaluation section of the site.

**Acceptance criteria:**
- [x] When a user with an evaluator role logs in, they are sent to the Evaluator's dashboard on the Rails app, #52 
- [x] An appropriate utility menu is displayed for the user, #27 
- [x] If any other user who does not have an evaluator's role assigned to them tries to navigate to the URL directly (access evaluator role specific features), they are redirected to their `/dashboard` (based on their assigned role) ~and an error message is displayed: "You do not have access to this section of Challenge.gov."~

**Development tasks:**

- [x] #158

### Definition of Done

**Doing (dev team)**
- [x] Code complete 
- [x] Code is organized appropriately
- [x] Any known trade offs are documented in the associated GH issue
- [x] Code is documented, modules, shared functions, etc.
- [x] Automated testing has been added or updated in response to changes in this PR
- [x] The feature is smoke tested to confirm it meets requirements
- [x] Database changes have been peer reviewed for index changes and performance bottlenecks
- [ ] ~PR that changes or adds UI~
  - [ ] ~include a screenshot of the WAVE report for the altered pages~
  - [ ] ~Confirm changes were validated for mobile responsiveness~
- [x] PR approved / Peer reviewed 
- [x] Security scans passed
- [x] Automate accessibility tests passed
- [x] Build process and deployment is automated and repeatable
- [x] Feature toggles if appropriate
- [x] Deploy to staging
- [x] Move card to testing column in the board
---
**Staging**
- [x] Usability testing: mobile and desktop (Tracy or Marni)
- [ ] ~Accessibility tested (Marni)~
  - [ ]  Keyboard navigation
  - [ ]  Focus confirmed
  - [ ] Color contrast compliance
  - [ ] Screen reader testing
- [ ] ~Cross browser testing - UI rendering is performant on below listed devices/browsers (Tracy or Marni)~
  - [ ] Windows/Chrome
  - [ ] Windows/Edge
  - [ ] Mac/Chrome
  - [ ] Mac/Safari
  - [ ] iOS/Safari
- [x] AC review (Renata)
- [ ] ~Deploy to production (production-like environment for eval capability) (dev team)~
- [x] Move to production column in the board
---
**Production**
- [ ] User and security documentation has been reviewed for necessary updates (Renata/Tracy/Dev team) 
- [x] PO / PM approved (Jarah or Renata)
- [x] AC is met and it works as expected (Jarah or Renata)
- [ ] Move to done column in the board (Jarah or Renata)

### LINKED ISSUE #23: Public solver role access
### User Story
As a user with a public solver role, in order to submit my solution to a challenge, I would like to be able to access appropriate sections of the site upon login.

**Acceptance criteria:**

- [x] There are no layout or functional changes for the public solver's role:
  - [x] When a user with a public solver role logs in, they are sent to the Public Solver's dashboard on the Phoenix app
  - [x] When the user selects "Apply for this challenge" on the challenge listing page, the system assumes that the user is public solver and they are routed to an appropriate app
  - [x] The public solver only has access to the Phoenix app
- [ ] **Will revisit if this is needed. Make a ticket** ~If any other user who does not have a public solver's role assigned to them tries to access public solver role specific features, they are redirected to their /dashboard (based on their assigned role) and an error message is displayed: "You do not have access to this section of Challenge.gov."~

**Development tasks:**

- [x] #158

### Definition of Done

**Doing (dev team)**
- [x] Code complete 
- [x] Code is organized appropriately
- [x] Any known trade offs are documented in the associated GH issue
- [x] Code is documented, modules, shared functions, etc.
- [x] Automated testing has been added or updated in response to changes in this PR
- [x] The feature is smoke tested to confirm it meets requirements
- [x] Database changes have been peer reviewed for index changes and performance bottlenecks
- [ ] ~PR that changes or adds UI~
  - [ ] ~include a screenshot of the WAVE report for the altered pages~
  - [ ] ~Confirm changes were validated for mobile responsiveness~
- [x] PR approved / Peer reviewed 
- [x] Security scans passed
- [x] Automate accessibility tests passed
- [x] Build process and deployment is automated and repeatable
- [x] Feature toggles if appropriate
- [x] Deploy to staging
- [x] Move card to testing column in the board
---
**Staging**
 - [ ] ~Accessibility tested (Marni)~
 - [ ]  ~Keyboard navigation~
 - [ ]  ~Focus confirmed~
 - [ ] ~Color contrast compliance~
- [ ] ~Screen reader testing~
- [x] Usability testing: mobile and desktop (Tracy or Marni)
- [x] Cross browser testing - UI rendering is performant on below listed devices/browsers (Tracy or Marni)
  - [x] Windows/Chrome
  - [x] Windows/Edge
  - [x] Mac/Chrome
  - [x] Mac/Safari
  - [x] iOS/Safari
- [x] AC review (Renata)
- [ ] ~Deploy to production (production-like environment for eval capability) (dev team)~
- [x] Move to production column in the board
---
**Production**
- [ ] User and security documentation has been reviewed for necessary updates (Renata/Tracy/Dev team) 
- [x] PO / PM approved (Jarah or Renata)
- [x] AC is met and it works as expected (Jarah or Renata)
- [ ] Move to done column in the board (Jarah or Renata)

### LINKED ISSUE #158: Add role-based permission check to routes
_As a admin of the system, I want to ensure that users only access parts of the application that are appropriate to their role._

- [x] Add role-based permission checks to the pages that require specific roles:
  - [x] `/evaluation_forms` can only be accessed by `Challenge Manager` role
  - [x] `/manage_submissions` can only be access by `Challenge Manager` role
  - [x] `/evaluations` can only be access by `Evaluator` role
- [x] Redirect users to their respective dashboards
  - [x] `Solver` role is redirected to the phoenix app if they try to access any of the above features
- [x] Show the flash message "You do not have access to this task. Please select a different task, or contact Challenge.gov support team for assistance." after they're redirected

### PR DESCRIPTION
**Pull request by cpreisinger**
Main tasks
#158 

Role specific sub tasks
#13 
#16 
#23

### CODE REVIEW

File: .codeclimate.yml [hunk 1]
```diff
@@ -15,4 +15,5 @@ exclude_patterns:
 - db/
 - vendor/
 -  "**/vendor/**/*"
-- app/assets/images/
\ No newline at end of file
+- app/assets/images/
+- spec/**/*
```
Comment (stepchud) [inline .codeclimate.yml:19]: 👍 nice, even codeclimate recommends to [exclude specs/tests](https://docs.codeclimate.com/docs/excluding-files-and-folders#auto-generated-file-and-folder-exclusions) from the quality check

File: app/models/user.rb [hunk 2]
```diff
@@ -82,6 +82,9 @@ class User < ApplicationRecord
 
   validates :email, presence: true
 
+  ROLES = %w[challenge_manager evaluator solver].freeze
+  validates :role, inclusion: { in: ROLES }
```
Comment (stepchud) [inline app/models/user.rb]: on the phx side there are a few admin roles too, shouldn't we include those roles too? atm I'm not sure what those roles can do in the Rails app yet.

File: spec/requests/dashboard_request_spec.rb [hunk 3]
```diff
@@ -15,8 +15,7 @@
         get "/"
       end
 
-      it_behaves_like "a page with utility menu links for all users"
```
Comment (stepchud) [inline spec/requests/dashboard_request_spec.rb:18]: shouldn't behavior this be kept?

File: spec/requests/evaluation_forms_spec.rb [hunk 4]
```diff
@@ -2,10 +2,41 @@
...
+        expect(response).to redirect_to(dashboard_path)
+      end
+    end
+
+    context "when logged in as a solver" do
+      before do
+        create_and_log_in_user(role: "solver")
+      end
+
+      it "renders the index view with the correct header" do
```
Comment (stepchud) [inline spec/requests/evaluation_forms_spec.rb]:
```suggestion
      it "redirects to the phoenix app" do
```

File: spec/requests/evaluation_forms_spec.rb [hunk 5]
```diff
@@ -2,10 +2,41 @@
...
+        expect(response.body).to include("Evaluation Forms")
+      end
+    end
+
+    context "when logged in as an evaluator" do
+      before do
+        create_and_log_in_user(role: "evaluator")
+      end
+
+      it "renders the index view with the correct header" do
```
Comment (stepchud) [inline spec/requests/evaluation_forms_spec.rb]:
```suggestion
      it "redirects to the dashboard" do
```

File: spec/requests/evaluations_spec.rb [hunk 6]
```diff
@@ -2,10 +2,41 @@
...
-    it "returns http success" do
-      get "/evaluations"
-      expect(response).to have_http_status(:success)
-      expect(response.body).to include("Evaluations")
+    context "when logged in as a challenge manager" do
+      before do
+        create_and_log_in_user(role: "challenge_manager")
+      end
+
+      it "renders the index view with the correct header" do
```
Comment (stepchud) [inline spec/requests/evaluations_spec.rb]:
```suggestion
      it "redirects to the dashboard" do
```

File: spec/requests/evaluations_spec.rb [hunk 7]
```diff
@@ -2,10 +2,41 @@
...
+        expect(response.body).to include("Evaluations")
+      end
+    end
+
+    context "when logged in as a solver" do
+      before do
+        create_and_log_in_user(role: "solver")
+      end
+
+      it "renders the index view with the correct header" do
```
Comment (stepchud) [inline spec/requests/evaluations_spec.rb]:
```suggestion
      it "redirects to the phoenix app" do
```

File: spec/requests/manage_submissions_spec.rb [hunk 8]
```diff
@@ -2,10 +2,41 @@
...
+        expect(response.body).to include("Manage Submissions")
+      end
+    end
+
+    context "when logged in as an evaluator" do
+      before do
+        create_and_log_in_user(role: "evaluator")
+      end
+
+      it "renders the index view with the correct header" do
```
Comment (stepchud) [inline spec/requests/manage_submissions_spec.rb]:
```suggestion
      it "redirects to the dashboard" do
```

File: spec/requests/manage_submissions_spec.rb [hunk 9]
```diff
@@ -2,10 +2,41 @@
...
+        expect(response).to redirect_to(dashboard_path)
+      end
+    end
+
+    context "when logged in as a solver" do
+      before do
+        create_and_log_in_user(role: "solver")
+      end
+
+      it "renders the index view with the correct header" do
```
Comment (stepchud) [inline spec/requests/manage_submissions_spec.rb]:
```suggestion
      it "redirects to the phoenix app" do
```

File: spec/requests/evaluations_spec.rb (hunk 6 above)
Comment (stepchud) [reply spec/requests/evaluations_spec.rb]: can we assert the error flash message for these too?

**Review by stepchud — CHANGES_REQUESTED** [review]
some feedback mostly on the specs

File: spec/requests/dashboard_request_spec.rb (hunk 3 above)
Comment (cpreisinger) [reply spec/requests/dashboard_request_spec.rb:18]: Solvers go to an entirely different app so it wouldn't be showing the rails utility menu right? That's why I was wondering if the solver utility menu wasn't needed. Or at least not until that is moved over

Comment (stepchud) [reply spec/requests/dashboard_request_spec.rb:18]: oh my bad, i didn't see that was for public solvers, carry on 👌

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -25,6 +25,9 @@ def log_in_user(user)
 end
 
 def create_and_log_in_user(user_attrs = {})
+  defaults = { email: "test@example.com", role: "challenge_manager" }
+  user_attrs = defaults.merge(user_attrs)
```
Comment (stepchud) [inline spec/spec_helper.rb]: can you move these defaults into the `create_user` function, since that's ultimately where they take affect

**Review by stepchud — APPROVED** [review]
LGTM 👍

**Pull request merged**